### PR TITLE
feat: add hasRoute

### LINKED ISSUE #4215: feature to check if route already exists
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the feature has not already been requested

### 🚀 Feature Proposal

It would be very useful if the Fastify instance exposes a method to check if a route is already registered in the current context just like it is implemented for content type parsers and constraints.

### Motivation

This feature could be useful for plugins to check if a certain route already exists and then add the route only if it does not yet exist in the current context.

### Example

```js
const fastify = require("fastify");

const app = fastify()

app.register(async (instance) => {
  instance.get('/test', async () => { return 'firstInstance' })

  instance.register(async (secondInstance) => {
    if (!secondInstance.hasRoute('/test')) {
      secondInstance.get('/test', async () => { return 'secondInstance' }) 
    }
  })
})

app.ready()
```

### PR DESCRIPTION
**Pull request by Uzlopak**
Fixes #4215 

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by Uzlopak** [comment]
In which file belongs the documentation of this method?

**Comment by mcollina** [comment]
https://github.com/fastify/fastify/blob/main/docs/Reference/Server.md

**Comment by Uzlopak** [comment]
added documentation

**Review by mcollina — APPROVED** [review]
lgtm

File: docs/Reference/Server.md [hunk 1]
```diff
@@ -1048,6 +1049,20 @@ fastify.routing(req, res)
...
+
+Method to check if a route is already registered to the internal router:
+
+```js
+fastify.hasRoute({
+  url: '/',
+  method: 'GET',
+  constraints: { version: '1.0.0' }
+})
+```
```
Comment (jsumners) [inline docs/Reference/Server.md:1063]: What does this return? How is it used? What are the available properties of the input object? What is each of their purpose?

See lines 998-1003 above.

**Review by jsumners — CHANGES_REQUESTED** [review]

File: docs/Reference/Server.md (hunk 1 above)
Comment (Uzlopak) [reply docs/Reference/Server.md]: better?

File: docs/Reference/Server.md [hunk 2]
```diff
@@ -1048,6 +1049,22 @@ fastify.routing(req, res)
...
+```js
+fastify.hasRoute({
+  url: '/',
+  method: 'GET',
+  constraints: { version: '1.0.0' } // optional
+})
+// true if route is already registered
+// false if route is not registered
+```
+
```
Comment (Uzlopak) [inline docs/Reference/Server.md:1066]:
```suggestion
```

File: docs/Reference/Server.md [hunk 3]
```diff
@@ -1048,6 +1049,21 @@ fastify.routing(req, res)
...
+Method to check if a route is already registered to the internal router:
+
+```js
+fastify.hasRoute({
+  url: '/',
+  method: 'GET',
+  constraints: { version: '1.0.0' } // optional
+})
+// true if route is already registered
+// false if route is not registered
```
Comment (jsumners) [inline docs/Reference/Server.md:1064]:
```suggestion
```js
const routeExists = fastify.hasRoute({
  url: '/',
  method: 'GET',
  constraints: { version: '1.0.0' } // optional
})

if (routeExists === false) {
  // add route
}
```

File: docs/Reference/Server.md [hunk 4]
```diff
@@ -1048,6 +1049,21 @@ fastify.routing(req, res)
 Method to add routes to the server, it also has shorthand functions, check
 [here](./Routes.md).
 
+#### hasRoute
+<a id="hasRoute"></a>
+
+Method to check if a route is already registered to the internal router:
```
Comment (jsumners) [inline docs/Reference/Server.md:1055]: I don't know what happened to my other comment, but these docs do not describe the required input object. What properties are possible? Which ones are required? The method signature should be fully documented.

**Review by jsumners — CHANGES_REQUESTED** [review]

File: docs/Reference/Server.md (hunk 4 above)
Comment (Uzlopak) [reply docs/Reference/Server.md:1055]: Do you have an example in the docs, so that i can do it with the same style?

Comment (Uzlopak) [reply docs/Reference/Server.md]: @jsumners 
better?

**Review by jsumners — APPROVED** [review]
Looks good to me.

**Pull request merged**